cmd/lncli/walletrpc_active.go

```diff
@@ -396,3 +401,200 @@ func labelTransaction(ctx *cli.Context) error {
+		return cli.ShowCommandHelp(ctx, "fundpsbt")
+	}
+
+	walletClient, cleanUp := getWalletClient(ctx)
+	defer cleanUp()
+
+	args := ctx.Args()
+	psbtBytes, err := base64.StdEncoding.DecodeString(args.First())
+	if err != nil {
+		return err
+	}
+
+	req := &walletrpc.FundPsbtRequest{
+		Template: &walletrpc.FundPsbtRequest_Psbt{
+			Psbt: psbtBytes,
+		},
+	}
+
+	switch {
+	case ctx.Uint64("conf_target") > 0 && ctx.Uint64("sat_per_vbyte") > 0:
+		return fmt.Errorf("cannot set conf_target and sat_per_vbyte " +
+			"at the same time")
+
+	case ctx.Uint64("conf_target") > 0:
+		req.Fees = &walletrpc.FundPsbtRequest_TargetConf{
+			TargetConf: uint32(ctx.Uint64("conf_target")),
+		}
+
+	case ctx.Uint64("sat_per_vbyte") > 0:
+		req.Fees = &walletrpc.FundPsbtRequest_SatPerVbyte{
+			SatPerVbyte: uint32(ctx.Uint64("sat_per_vbyte")),
+		}
+	}
+
+	response, err := walletClient.FundPsbt(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	printJSON(&fundPsbtResponse{
+		Psbt:  base64.StdEncoding.EncodeToString(response.FundedPsbt),
+		Locks: response.LockedUtxos,
+	})
+
+	return nil
+}
+
+// finalizePsbtResponse is a struct that contains JSOn annotations for nice
+// result serialization.
+type finalizePsbtResponse struct {
+	Psbt    string `json:"psbt"`
+	FinalTx string `json:"final_tx"`
+}
+
+var finalizePsbtCommand = cli.Command{
+	Name:      "finalizepsbt",
+	Category:  "On-chain",
+	Usage:     "Finalize a Partially Signed Bitcoin Transaction (PSBT).",
+	ArgsUsage: "psbt",
+	Description: `
+	The finalizepsbt command expects a partial transaction with all inputs
+	and outputs fully declared and tries to sign all inputs that belong to
+	the wallet. Lnd must be the last signer of the transaction. That means,
+	if there are any unsigned non-witness inputs or inputs without UTXO
+	information attached or inputs without witness data that do not belong
+	to lnd's wallet, this method will fail. If no error is returned, the
+	PSBT is ready to be extracted and the final TX within to be broadcast.
+
+	This method does NOT publish the transaction after it's been finalized
+	successfully.
+	`,
+	Action: actionDecorator(finalizePsbt),
+}
+
+func finalizePsbt(ctx *cli.Context) error {
+	// Display the command's help message if we do not have the expected
+	// number of arguments/flags.
+	if ctx.NArg() != 1 {
+		return cli.ShowCommandHelp(ctx, "finalizepsbt")
+	}
+
+	walletClient, cleanUp := getWalletClient(ctx)
+	defer cleanUp()
+
+	args := ctx.Args()
+	psbtBytes, err := base64.StdEncoding.DecodeString(args.First())
+	if err != nil {
+		return err
+	}
+
+	req := &walletrpc.FinalizePsbtRequest{
+		FundedPsbt: psbtBytes,
+	}
+	response, err := walletClient.FinalizePsbt(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	printJSON(&finalizePsbtResponse{
+		Psbt:    base64.StdEncoding.EncodeToString(response.SignedPsbt),
+		FinalTx: hex.EncodeToString(response.RawFinalTx),
+	})
+
+	return nil
+}
+
+var releaseOutputCommand = cli.Command{
```

Comment:
Hmm, alternatively, we could return a unique identifier for the PSBT packet (could just be the full txid) then have the user pass that in to somethign like `cancelpsbt` which looks up the transaction and does all the input unlocking automatically.

Reply:
But wouldn't we need to store the PSBT in the wallet then? The way it's currently implemented, the PSBT operations are stateless (except for an eventual change address being added to the wallet DB).

Reply:
Yeah that's true, it's nice that things are stateless as is.